src/test/ui/consts/miri_unleashed/const_refers_to_static.rs

```diff
@@ -1,8 +1,8 @@
 // build-fail
-// compile-flags: -Zunleash-the-miri-inside-of-you
+// compile-flags: -Zunleash-the-miri-inside-of-you -Zdeduplicate-diagnostics
 #![allow(const_err)]
-
-#![feature(const_raw_ptr_deref)]
+#![feature(const_raw_ptr_deref)] // FIXME: cannot remove because then rustc thinks there is no error
```

Comment:
This is so weird. The test passes (so rustc fails), but if I remove the flag I am hitting the new panic *and the errors are not even printed*. Do we drop the session before codegen, or so?!?

Reply:
hmm... maybe that panic causes a double panic when dropping the fields of `Session` during unwinding?

Maybe we should report this error not via a panic in a `Session` destructor, but wherever "n previous errors" is reported, because there should be no further things happening after that.

Reply:
I think my choice of using `Session::drop` may have been very much not thought through

Reply:
So, while https://github.com/rust-lang/rust/blob/86c0fca44bc0b49085b177d0244bc300f6cd4655/src/librustc_interface/interface.rs#L196 is the right point in time, I don't like putting it right there

Reply:
I don't see any other way though. So...

1. make the code from `Session::drop` a `Session::check_unleash_gates` (or choose a better name)
2. Create `Session::finish_diagnostics` which invokes the `compiler.sess.diagnostic().print_error_count(registry);` and invoke `finish_diagnostics` in https://github.com/rust-lang/rust/blob/86c0fca44bc0b49085b177d0244bc300f6cd4655/src/librustc_interface/interface.rs#L196 instead of doing the `print_error_count` directly there (I think you can get the `Registry` needed from the `Session`).
3. call `Session::check_unleash_gates` before the `print_error_count` in `finish_diagnostics`
4. Change the `panic!` to `self.err` invocation.

Reply:
Yes that seems to work. :) Commit incoming.